pkg/plugin/plugin.go

```diff
@@ -99,24 +102,37 @@ func (p *plugin) GenerateConfig(ctx context.Context, cs *api.OpenShiftManagedClu
 
 func (p *plugin) GenerateARM(ctx context.Context, cs *api.OpenShiftManagedCluster, isUpdate bool) ([]byte, error) {
 	log.Info("generating arm templates")
-	generator := arm.NewSimpleGenerator(p.entry)
-	return generator.Generate(ctx, cs, isUpdate)
+	return p.armGenerator.Generate(ctx, cs, isUpdate)
 }
 
 func (p *plugin) InitializeCluster(ctx context.Context, cs *api.OpenShiftManagedCluster) error {
 	log.Info("initializing cluster")
-	initializer := initialize.NewSimpleInitializer(p.entry, p.config)
-	return initializer.InitializeCluster(ctx, cs)
+	return p.clusterUpgrader.InitializeCluster(ctx, cs)
 }
 
 func (p *plugin) HealthCheck(ctx context.Context, cs *api.OpenShiftManagedCluster) error {
 	log.Info("starting health check")
-	healthChecker := healthcheck.NewSimpleHealthChecker(p.entry, p.config)
-	return healthChecker.HealthCheck(ctx, cs)
+	return p.clusterUpgrader.HealthCheck(ctx, cs)
 }
 
-func (p *plugin) Update(ctx context.Context, cs *api.OpenShiftManagedCluster, azuredeploy []byte) error {
-	log.Info("starting update")
-	upgrader := upgrade.NewSimpleUpgrader(p.entry, p.config)
-	return upgrader.Update(ctx, cs, azuredeploy, p.config)
+func (p *plugin) CreateOrUpdate(ctx context.Context, cs *api.OpenShiftManagedCluster, azuredeploy []byte, isUpdate bool, deployFn api.DeployFn) error {
```

Comment:
Hi @amanohar 

What error format do you need?
If we can't get the error format to your liking, I could make those functions public again.
We were trying to make improve api (there was some inconsistencies between what deploy and update were doing that you may not have noticed since you didn't use them), so if it is not working for you I can sort it out. 

I assume you want to go back to:
(Plugin) GenerateARM
RP deploys template, handles errors
(Plugin) InitializeCluster
(Plugin) WaitForEverythingToComeUp (was healthcheck)

re: Side bar
HealthCheck doesn't call waitForNodes and waitForInfraServices any more so there shouldn't be any duplicated calls.
https://github.com/openshift/openshift-azure/blob/master/pkg/upgrade/healthcheck.go#L18

Reply:
Hey @asalkeld, it not as much about a particular kind of error format. RP does more than that. Based on success or failure of template deployment, RP gets all deployment results that were returned by ARM (by making another call) and parses all errors to determine if error should be QoS impacting or not.  Also based on the failure it determines what was the impacting downstream RP and logs it. 

Sometimes RP changes the error code that is reported to the customer because an error code returned by CRP for e.g. might not be relevant to OSA customer. Plus deploy function does a lot of other things like retries. and is constantly being improved. 

The template deployment code also has preflight error handling baked into it that we will enable for OSA at some point. For all of this to happen properly RP needs the flexibility to be able to get a template and deploy it.

Most importantly, running create or update cannot be a black box function to the RP because each step in the process in unique & functionally different and putting it all in one function and returning it as a generic error reduces flexibility of the process overall & takes away visibility into how each error needs to be handled, logged, and reported from RP's side which is essential.

Reply:
@asalkeld, I must have missed that there are no dup calls amidst all the scrolling and diffing. Sorry about that. I will reread the code.